### PR TITLE
Added write dataset to ts file functionality

### PR DESCRIPTION
**Pull request by whackteachers**
#### Reference Issues/PRs

This is partly addressed to #434 which is a functionality requested by @TonyBagnall 

#### What does this implement/fix? Explain your changes.

I have implmented a new function that can writes a dataframe to a .ts file which is very useful for saving transformed dataset,
along with some unit testing.
#### Does your contribution introduce a new dependency? If yes, which one?
N/A

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/master/.all-contributorsrc).
- [ ] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/master/CODEOWNERS) to receive notifications about future changes to these files.
- [ ] I've added unit tests and made sure they pass locally.

##### For new estimators
- [ ] I've added the estimator to the online documentation.
- [ ] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.

#### Any other comments?

Just a side note, the function right now does not support timestamp data, as I have not seen an example of it in the dataset nor in the docs. I have also mention this in the doc string.

### CODE REVIEW

File: sktime/utils/validation/tests/test_write_dataframe_to_ts.py
Comment (mloning) [inline sktime/utils/validation/tests/test_write_dataframe_to_ts.py:9]: I don't think we need this import, pytest automatically recognises `tmp_path` as a test fixture see https://docs.pytest.org/en/stable/tmpdir.html

File: sktime/utils/results_writing.py
Comment (mloning) [inline sktime/utils/results_writing.py:88]: Regarding the restructuring of utils, this function should probably go into `data_container` and not `results_writing`, no?

Comment (mloning) [reply sktime/utils/results_writing.py:88]: I'd also set a default argument for `problem_name` so that users don't have to specify it whenever they just want to write out some data

File: sktime/utils/results_writing.py
Comment (mloning) [inline sktime/utils/results_writing.py:133]: some superfluous horizontal lines here

**Review by mloning — COMMENTED** [review]
Looks generally good to me, very clean and readable code, thanks @whackteachers!

File: sktime/utils/results_writing.py
Comment (TonyBagnall) [reply sktime/utils/results_writing.py:88]: hi markus, I asked him to put it here for now, we are delaying all refactoring until you give us the green light 👍

**Comment by TonyBagnall** [comment]
I think the code quality issue is just some of my code in experiments.py and not relevant to this PR. The travis issue may be caused by the fact that @whackteachers does not have travis installed on his fork? We encountered something similar with tsml.

**Comment by mloning** [comment]
The travis build didn't error, just stalled, I re-started and should pass now. 

To fix the code-quality check, see this post: https://gist.github.com/mloning/6f4b8d412743968af6b2da8b4eaaf2bf

**Comment by whackteachers** [comment]
> The travis build didn't error, just stalled, I re-started and should pass now.
> 
> To fix the code-quality check, see this post: https://gist.github.com/mloning/6f4b8d412743968af6b2da8b4eaaf2bf

I have ran pre-commit before I push and it was fine, also the command given is not working, give smth like --name-only master is not a valid parameter error

**Comment by mloning** [comment]
Okay happy to ignore this, since it's only on the contrib/experiments.py file

**Review by TonyBagnall — DISMISSED** [review]
looks good to me

**Comment by mloning** [comment]
Hi @whackteachers, I just merged another big PR which created a merge conflict here, once you resolve it, we can merge this PR too. Let me know if you need help with that!

File: sktime/utils/results_writing.py
Comment (mloning) [inline sktime/utils/results_writing.py:114]: Does path not include the file name? Below it says that `problem_name` gives the file name. So does this give the directory?

File: sktime/utils/results_writing.py
Comment (mloning) [inline sktime/utils/results_writing.py:180]: Could we use the input validation functions from `sktime.utils.validation.panel`, i.e. `check_X` with `coerce_to_pandas=True` so that this also works when passing a 3d numpy array?

File: sktime/utils/validation/tests/test_write_dataframe_to_ts.py
Comment (mloning) [inline sktime/utils/validation/tests/test_write_dataframe_to_ts.py:42]: Why is this called `ItalyPowerDemand_transform.ts`, where does the `transform` come from?

**Review by mloning — COMMENTED** [review]
Three minor comments, everything looks good!

File: sktime/utils/results_writing.py
Comment (whackteachers) [reply sktime/utils/results_writing.py:114]: no, path just refers to the directory to store the file, while problem_name is used to name the ts file at the same time.

File: sktime/utils/results_writing.py
Comment (whackteachers) [reply sktime/utils/results_writing.py:180]: yeah sure, I wasn't aware there is already such module exist

File: sktime/utils/validation/tests/test_write_dataframe_to_ts.py
Comment (whackteachers) [reply sktime/utils/validation/tests/test_write_dataframe_to_ts.py:42]: the write_dataframe_to_tsfile appends _transformat the end of the problem name as the file name, see
https://github.com/whackteachers/sktime/blob/3591a48577a7cda6fb2b5fffa8cdf1f6597107c2/sktime/utils/results_writing.py#L194-L201

Comment (TonyBagnall) [reply sktime/utils/validation/tests/test_write_dataframe_to_ts.py:42]: hi, this is fine, but fyi general format for storing results we use is 
<problem name>/<problem_name>.ts
If I were doing a transform on multiple data I would either have a whole directory and keep the original name
<transform name>/<problem name>/<problem_name>.ts
or append the transform name to all
<problem name_transform name>/<problem_name_transform name>.ts

**Review by TonyBagnall — APPROVED** [review]

File: sktime/utils/validation/tests/test_write_dataframe_to_ts.py
Comment (whackteachers) [reply sktime/utils/validation/tests/test_write_dataframe_to_ts.py:42]: ok thanks for the reminder, I will keep this in mind next time

File: sktime/utils/results_writing.py
Comment (mloning) [reply sktime/utils/results_writing.py:114]: Okay I think this is worth clarifying in the doc strings, usually input args are data and path including file name (e.g. `np.savetxt` or `pd.to_csv`)

File: sktime/utils/validation/tests/test_write_dataframe_to_ts.py
Comment (mloning) [reply sktime/utils/validation/tests/test_write_dataframe_to_ts.py:42]: from a user perspective, I find it strange if I give a path (directory) and problem name but then can't find the generated file, no? would probably prefer to have a user friendly function which takes data and file (path with file name, e.g. "my_data.ts") and a wrapper for your benchmarking purposes

Comment (TonyBagnall) [reply sktime/utils/validation/tests/test_write_dataframe_to_ts.py:42]: one reason for this is that we need to organise multiple problems, each of which has 6 or more files associated with it. These may be predefined splits into train and test or resamples. This structure allows us to automate much of the internal data complexity without bothering the user. For example, if you load a problem it looks to detect if there is already an predefined train test split, and it also looks to see if the resamples are already generated.

Comment (mloning) [reply sktime/utils/validation/tests/test_write_dataframe_to_ts.py:42]: yes I think that makes sense, but can that not be handled by a wrapper which defines additional behaviour for your purposes, so that we can still have a more generic function for writing a data frame to a ts file for users, i.e. the counterpart to `load_from_tsfile_to_dataframe`

Comment (TonyBagnall) [reply sktime/utils/validation/tests/test_write_dataframe_to_ts.py:42]: Ultimately, sure, but that's beyond the scope of this PR

**Comment by mloning** [comment]
Alright, agree, let's merge this PR first and work on the more generic function in a separate PR. 

One of the travis build stalled again, just restarted it now and will merge later today once it passes. 

Thanks again @whackteachers - great work!

**Pull request merged**
